### PR TITLE
Poc for unit testing D3 bindings

### PR DESCRIPTION
**Pull request by giraud**
I added the structure needed to test the bindings as unit tests with jsdom before doing some changes.

### CODE REVIEW

**Comment by giraud** [comment]
oh ! I failed the PR. I wanted to have https://github.com/af/bs-d3/pull/4/commits/d81c5f2b3bc64ad0014d7de4df0259a9a3a124f4 in a separate commit.
This commit try to make the code closer to D3js, for ex by removing the subSelect external. I took inspiration from bs-webapi and I manage namespace manually.
Tell me if you want me to create 2 separate commits.

File: src/SelectionD3.re [hunk 1]
```diff
@@ -0,0 +1,28 @@
+type t;
```
Comment (af) [inline src/SelectionD3.re:1]: What's the reason for having both this file and `Selection.re`?

**Comment by af** [comment]
Thanks for this! I do think that commit should be separate though, and let's keep the namespacing in for now. Back when I was working on this, chenglou recommended enabling it, though I'm not sure if that's still a best practice (there's some [current discussion](https://github.com/BuckleScript/bucklescript/issues/2688) on it)

**Comment by af** [comment]
Also thanks for fixing a few of the bindings that were completely wrong 😂

File: src/SelectionD3.re (hunk 1 above)
Comment (giraud) [reply src/SelectionD3.re:1]: I forgot to delete Selection.re

**Comment by giraud** [comment]
Hi, sure I'll create 2 commits (maybe in a week).

I totally agree with you on namespacing: it is very important for this kind of library to use namespacing.

As you can see in the PR, all files end with D3, like SelectionD3 or ArrayD3. 

So I am using namespacing, but manual namespacing (https://blog.janestreet.com/better-namespaces-through-module-aliases/). I am not using the **automatic** namespacing process from bsb, because it has limitations that prevent this library to have a clean API. By clean API I mean an API that mimic the one from D3.

Currently, you need to do that: `D3.Selection.select("body") |> D3.Selection.subSelect("p")`.
With my proposal you do: `D3.select("body") |> D3.Selection.select("p")`.
Both are converted to the same js: `d3.select("body").select("p")`.

diff:
- subSelect doesn't exist in d3 and will confuse developers
- the root selection `D3.select(...)` is closer to the js code `d3.select(...)`

That's why I think hat manual namespacing is the way to go right now, until we can do the same with automatic namespacing.  
But maybe in the next PR I should use D3_ instead to make it more clear (D3_Selection, D3_Array) ?

wdyt?

**Comment by af** [comment]
I do like your selection API better, just not sure if we're giving up anything by doing namespacing manually instead of using bsb's built-in support. I'll try and learn more about the latter.

**Comment by giraud** [comment]
Hi, did you find an answer ?

**Comment by af** [comment]
@giraud Sorry, this kind of fell off the map for me! If you could remove the redundant Selection.re file, I'll take another look at getting this merged. Also could you move the testing scaffolding stuff out of this PR, and introduce them in a separate PR where we can iterate on testing?

Also for the record, I don't think we will be able to follow d3's API exactly. To provide the same level of flexibility in a typed language, there will probably need to be lots of functions on top of d3's surface area (eg for d3js functions that overload functions and support multiple typed arguments).

**Comment by giraud** [comment]
I'll do it, I just wanted to be sure there is no blocker before working on it again.

**Pull request closed without merging**